turbine/src/cluster_nodes.rs

```diff
@@ -328,6 +334,30 @@ fn get_nodes(cluster_info: &ClusterInfo, stakes: &HashMap<Pubkey, u64>) -> Vec<N
     // Since sorted_by_key is stable, in case of duplicates, this
     // will keep nodes with contact-info.
     .dedup_by(|a, b| a.pubkey() == b.pubkey())
+    .filter_map(|node| {
+        if !should_dedup_addrs
+            || node
+                .contact_info()
+                .and_then(|node| node.tvu(Protocol::UDP).ok())
+                .map(|addr| {
+                    *counts
+                        .entry(addr.ip())
+                        .and_modify(|count| *count += 1)
+                        .or_insert(1)
+                })
+                <= Some(MAX_NUM_NODES_PER_IP_ADDRESS)
+        {
+            Some(node)
+        } else {
+            // If the node is not staked, drop it entirely. Otherwise, keep the
+            // pubkey for deterministic shuffle, but strip the contact-info so
+            // that no more packets are sent to this node.
+            (node.stake > 0u64).then(|| Node {
```

Comment:
The random shuffle for "staked" nodes should be deterministic across cluster and it cannot rely on gossip.
So we cannot drop staked nodes here because that depends on what other contact-infos you have observed in gossip which is obviously not deterministic across nodes.

Reply:
ah! good point...